ibis/backends/base/connection.py

```diff
@@ -0,0 +1,187 @@
+    """
+    Base class for all backend connections.
+
+    All Ibis backends are expected to subclass this `Connection` class,
+    and implement all the required methods.
+    """
+
+    @property
+    @abc.abstractmethod
+    def name(self) -> str:
+        """
+        Name of the backend, for example 'sqlite'.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def kind(self) -> str:
+        """
+        Backend kind. One of:
+
+        sqlalchemy
+            Backends using a SQLAlchemy dialect.
+        sql
+            SQL based backends, not based on a SQLAlchemy dialect.
+        pandas
+            Backends using pandas to store data and perform computations.
+        spark
+            Spark based backends.
+        """
+        pass
+
+    @abc.abstractmethod
+    def connect(connection_string: str, **options) -> BaseConnection:
+        """
+        Connect to the underlying database and return a client object.
+        """
+        pass
+
+    def register_options(self) -> None:
+        """
+        If the backend has custom options, register them here.
+        They will be prefixed with the name of the backend.
+        """
+        pass
+
+    def compile(self, expr: ir.Expr, params=None) -> str:
+        """
+        Compile the expression.
+        """
+        return self.compiler.to_sql(expr, params=params)
+
+    def verify(self, expr: ir.Expr, params=None):
+        """
+        Verify `expr` is an expression that can be compiled.
+        """
+        try:
+            self.compile(expr, params=params)
+        except TranslationError:
+            return False
+        else:
+            return True
+
+    def add_operation(self, operation: Callable) -> Callable:
+        """
+        Decorator to add a translation function to the backend for a specific
+        operation.
+
+        Operations are defined in `ibis.expr.operations`, and a translation
+        function receives the translator object and an expression as
+        parameters, and returns a value depending on the backend. For example,
+        in SQL backends, a NullLiteral operation could be translated simply
+        with the string "NULL".
+
+        Examples
+        --------
+        >>> @ibis.sqlite.add_operation(ibis.expr.operations.NullLiteral)
+        ... def _null_literal(translator, expression):
+        ...     return 'NULL'
+        """
+        if not hasattr(self.client, 'compiler'):
+            raise RuntimeError(
+                'Only SQL-based backends support `add_operation`'
+            )
+
+        def decorator(translation_function):
+            self.client.compiler.translator_class.add_operation(
+                operation, translation_function
+            )
+
+        return decorator
+
+    @property
+    def current_database(self) -> str:
+        """Return the current database."""
+        # TODO: If we can assume this, we should define the `con` attribute
+        return self.con.database
```

Comment:
What is `self.con`?

Reply:
It's assumed to be the actual connection with the database, a DBAPI2 `Connection` instance. This is hacky, I added the TODO to take care of this later, but this is how it is right now in master, just copied the method.

Reply:
kk